### ISSUE TITLE
Add .env support, take 2

### ISSUE THREAD

**Issue by davidism** [pallets/flask#2375]
Setting the `FLASK_APP` and `FLASK_DEBUG` for every new terminal session is inconvenient. The CLI could load a `.env` file that contained these values instead. Since Heroku is a popular deployment option, this would also help developers who want the same scheme locally.

Instead of writing our own parser, we should use [python-dotenv](https://github.com/theskumar/python-dotenv). This reduces maintenance burden on us and can be an optional dependency. If it is installed, the CLI will load the `.env`, otherwise it will skip it.

Previously discussed in #2164.

---

@miguelgrinberg raised a good point in the previous thread. `.env` is typically used to store sensitive configuration. The convenience of setting `FLASK_APP=` there could encourage users to commit the file publicly, which would be bad.

A solution would be to load two files, `.flask-env` and `.env`, and prominently document that `.flask-env` can be committed but should not contain secrets.

**Comment by ThiefMaster** [comment]
i'd prefer `.flaskenv` similar to `.foorc` and not `.foo-rc`

**Comment by davidism** [comment]
That makes sense, I have no strong opinion on the name.

**Comment by miguelgrinberg** [comment]
I'm in favor of using python-dotenv. I have started using it on my projects, in place of my own less robust implementation. I'm also in favor of having Flask import a `.env` file, that is not supposed to be in source control. I think this would be a great addition regardless of the main point of this discussion.

I think using a `.flaskenv` for the Flask env vars is a better idea than what we discussed in #2164. But I keep thinking that for people that use app factory functions you now need to add two files: a `wsgi.py` that calls the factory function, and then this new `.flaskenv` that points to it.

Implementing support for referencing a factory function in `FLASK_APP` is, in my opinion, not a great idea because the function might need custom arguments. So I see no way to avoid the need for a `wsgi.py` file.

On the other side, the Flask CLI is perfectly happy to look for `app` or `application` in the module referenced by `FLASK_APP` if you don't give it a specific app name. So what if we also add sensible defaults for module name? For example, if `FLASK_APP` is not defined, we look for `wsgi.py` and `flaskapp.py` and use those. So then we have the following options:

- `FLASK_APP` is not defined: look for `app` or `application` in `wsgi.py` or `flaskapp.py`.
- `FLASK_APP=module.py`: look for `app` or `application` in`module.py`
- `FLASK_APP=:myapp`: look for `myapp` in `wsgi.py` or `flaskapp.py`

**Comment by davidism** [comment]
@miguelgrinberg at the PyCon sprints, @HndrkMkt merged in a set of patches that allows `flask run` to autodetect factories and `FLASK_APP` to specify them, optionally with arguments. So no wrapper file will be required in the next version. #2297, #2319, #2326

**Comment by davidism** [comment]
I'm not sure about autodetecting the module, that seems a bit too magical. (Yes, I realize this is weird given the rest of the autodetection.) It would only benefit single-file apps, and devs like to choose all sorts of file names whether their app is a module or a package. They would also still use `FLASK_DEBUG`.

Anyway, that's sort of tangential to the dotenv support. We should discuss that in a separate issue.

**Comment by miguelgrinberg** [comment]
In my opinion, `FLASK_APP` is not in the same category as `FLASK_DEBUG`. The former can be considered static and universal. Needs to be set regardless of environment (dev, prod or test). The latter is a configuration setting that only applies to specific cases. And it's a setting that you can also set via the Flask config object, or the `--debug` CLI option.

The reason I brought app the defaults for module names is that if we agree to implement them, then you have a way to not need to use `FLASK_APP`.

Regarding the factory support for `FLASK_APP`, it looks pretty good, but it does not seem to cover all the possible cases. Basically I would need to do something like this:

    FLASK_APP=app:create_app(os.getenv('FLASK_CONFIG') or 'default')

The `create_app` function is in the `app/__init__.py` module. I need to debug this, but Flask master fails to import `app` as given above, for some reason. If I move `create_app` to a top-level module then that part works, but the expression I need for the function's single argument does seem to be accepted, the factory function gets `script_info` as this configuration argument.

**Comment by HndrkMkt** [comment]
@miguelgrinberg, that it fails to evaluate your argument expression is due to the fact that @davidism and I decided to limit the argument support to literals. (See https://github.com/pallets/flask/pull/2326/files#diff-61bbd5ac9238364564a7315cc6bc2156R114 for the evaluation).

I'll prepare a PR later that updates docs and tests to point out this limitation.

**Comment by davidism** [comment]
Entering quotes in an env var is unfortunately not straightforward. We should document that. Also, it's possible to reference a var or default.

**Comment by davidism** [comment]
```bash
# need nested quotes to get "string" in arg list
FLASK_APP=hello:create_app("'dev'") flask run
# can get var or default
FLASK_CONFIG="'dev'" FLASK_APP=hello:create_app(${FLASK_CONFIG:-"'default'"}) flask run
```

Although in the second one I'd probably go with:

```python
def create_app(config=None):
    if config is None:
        config = os.environ.get('FLASK_CONFIG', 'default')
    ...
```

```bash
FLASK_APP=hello flask run
```

**Comment by miguelgrinberg** [comment]
Regarding quotes, looks like I did not do it above in my example, but for bash, the easiest is to enclose the whole thing in doble quotes, and only use single quotes for your literals. This is actually what I tried yesterday and failed to correctly copy here:

    FLASK_APP="app:create_app(os.getenv('FLASK_CONFIG') or 'default')"

But of course, this works in bash/zsh/etc only. I suspect you are going to have a lot more trouble to get something like this to work on Windows.

Your suggested solution of letting the shell do the env var replacement should work, but again, it is a bash only solution.

Moving this logic into the factory function, as you also suggested above, is not ideal. The default could be context based, so it is better handled by the caller. When you are running unit tests, for example, you will probably want the default configuration to be `'test'` instead of `'default'`. If you are starting a Celery worker for your app, for example, you'll probably want yet another configuration to be the default.

I would personally not go through all these troubles to define `FLASK_APP`. Adding a `wsgi.py` file (which is already covered in the docs) is way more simple, and allows you to work in Python, regardless of the OS and shell that you use. So what I'm wondering is if we can agree that having a default for module name thus making `FLASK_APP` optional makes sense or not, maybe in addition to the .flaskenv effort which you seem to prefer.

**Comment by davidism** [comment]
If the default changes per context, then it's not a default. Isn't that the whole point of changing the env var or the argument to the factory?

Anyway, I understand that the wrapper file can still be useful for complex situations. Recognizing `wsgi.py` as a standard wrapper name, and `app.py` to support quick demo apps, seems fine.

**Comment by davidism** [comment]
I've moved detecting `wsgi.py` to a separate issue, #2376. Discussion here should be about the logistics of `.env` support itself.

**Comment by davidism** [comment]
Yes, debug should be documented to go in env not flaskenv.

This also doesn't help the case for packaging your app, which I really like to do. Including the `.flaskenv` file in the sdist wouldn't accomplish anything, doubly so for bdist_wheel. I guess that's another good argument for `wsgi.py` as an entry point.

**Comment by bovarysme** [comment]
Repeating an issue that was brought up in the previous thread: if `.env` is used to store secrets and is only loaded when Flask's CLI is invoked, it might have a limited use on production environments that use a WSGI server to run the application.

On the contrary, if `.env` is loaded on app instanciation, does it deprecate instance relative configurations? It seems to me these features both solve the same problem (having different secrets on different environments). For example one could write multiple `.env` files (`.env.production`, `.env.staging`, `.env.development`), not checked in version control, and push the correct file on each environment using a configuration management tool.

**Comment by davidism** [comment]
Instance folders don't work on services like Heroku that place their configuration in env vars and don't have persistent storage. This helps mirror that locally for developers using those services. They still have to do extra work to convert env vars to proper config values. This shouldn't replace instance folders.

I see this as a convenience during development. In production, there's probably something else like Systemd or Docker setting env vars. Of course, if a developer still wants to use a `.env` file they can use python-dotenv manually during `create_app` or in `wsgi.py`.

**Comment by miguelgrinberg** [comment]
I am going to side with @bovarysme. If the `.env` file is going to be imported, it needs to be imported regardless of how the application is started. It's too confusing and error prone to have this work only when you run the application via the flask cli.

I agree that it is a feature designed to help development, but what if I want to do a quick test running gunicorn or uwsgi? Maybe I won't even realize that I'm running a completely different configuration.

**Comment by davidism** [comment]
Where should it be loaded then? `Flask.__init__`? It needs to be loaded before that for the CLI, but shouldn't hurt to load it in both places.

**Comment by miguelgrinberg** [comment]
Right, that sounds tricky. Would it make sense to do it in `flask/__init__.py`? This is the main `flask` package, not the `Flask` class.

**Comment by davidism** [comment]
I'm not sure if I'm comfortable with import side effects, although I can't think why it would hurt in this case.

**Comment by justanr** [comment]
I think loading in any sort of init should be avoided if at all reasonably possible, using dealt with packages that do that it's an absolute nightmare, not just in testing but sometimes in just running the thing.

What value does this env file serve over a cfg file that's loaded at run time? And will this open the door to people asking for other config options - do we then debate loading from redis or consul? The answer is hopefully no, but this does set the precedent for considering those if accepted.

If this is accepted as something we want, would a new method on the config object be more appropriate? Say one that accepted an env file and how to handle it?

**Comment by ThiefMaster** [comment]
How about making it a normal function e.g. in `flask.util` and then call it from both the CLI code and from `Flask.__init__`?

**Comment by davidism** [comment]
The immediate problem this was proposed to solve was setting the env vars for the cli, which has to happen before the app is available. The discussion is around keeping the behavior when not using the cli.

Note that this will all be optional however it ends up working. If dotenv is not installed it will be skipped.

I think loading in `Flask.__init__` with an argument to disable would be acceptable. In addition to loading early in the cli.

**Comment by lepture** [comment]
> Setting the FLASK_APP and FLASK_DEBUG for every new terminal session is inconvenient.

Is this the problem that we should solve? `.env` may be a solution, but it is not the only solution. I prefer the flask-script way.

e.g. here is the `app.py`:

```python
app = create_app()

if __name__ == '__main__':
    app.create_cli()
```

And then, we can call `python app.py run`.

**Comment by davidism** [comment]
@lepture you can still do that:

```python
cli = FlaskGroup(create_app=create_app)

if __name__ == '__main__':
    cli()
```

You can also add an entry point for it:

```python
setup(
    ...,
    entry_points={
        'console_scripts': [
            'myapp=myapp:cli',
        ],
    },
)
```

```
myapp run
```

This still doesn't replace the `FLASK_DEBUG` env var though.
